aidial_adapter_bedrock/llm/converse/input.py

```diff
@@ -0,0 +1,293 @@
+                            }
+                        )
+        case None:
+            pass
+        case _:
+            assert_never(message.content)
+
+    if message.custom_content and message.custom_content.attachments:
+        for attachment in message.custom_content.attachments:
+            resource = await AttachmentResource(attachment=attachment).download(
+                storage
+            )
+            content.append(
+                {
+                    "image": {
+                        "format": to_converse_image_type(resource.type),
+                        "source": {"bytes": resource.data},
+                    }
+                }
+            )
+    if message.function_call and message.tool_calls:
+        raise ValidationError(
+            "You cannot use both function call and tool calls in the same message"
+        )
+    elif message.function_call:
+        content.append(function_call_to_content_part(message.function_call))
+    elif message.tool_calls:
+        content.extend(
+            [
+                tool_call_to_content_part(tool_call)
+                for tool_call in message.tool_calls
+            ]
+        )
+
+    return content
+
+
+async def to_converse_message(
+    message: DialMessage,
+    storage: FileStorage | None,
+    supported_image_types: list[str] | None = None,
+) -> ConverseMessage:
+
+    return {
+        "role": to_converse_role(message.role),
+        "content": await _get_converse_message_content(
+            message, storage, supported_image_types
+        ),
+    }
+
+
+def get_converse_system_prompt(
+    messages: List[DialMessage],
+) -> ConverseTextPart | None:
+    system_msgs = []
+    saw_non_system = False
+
+    for msg in messages:
+        if msg.role == DialRole.SYSTEM:
+            if saw_non_system:
+                raise ValidationError(
+                    "System messages are only allowed at the beginning of the messages list"
+                )
+            if isinstance(msg.content, str):
+                system_msgs.append(msg.content)
+        else:
+            saw_non_system = True
+
+    combined = "\n\n".join(msg for msg in system_msgs if msg)
+    return {"text": combined} if combined else None
+
+
+async def process_messages(
+    messages: List[DialMessage],
+    storage: FileStorage | None,
+) -> ListProjection[ConverseMessage]:
+    def _merge(
+        a: Tuple[ConverseMessage, Set[int]],
+        b: Tuple[ConverseMessage, Set[int]],
+    ) -> Tuple[ConverseMessage, Set[int]]:
+        (msg1, set1), (msg2, set2) = a, b
+
+        content1 = msg1["content"]
+        content2 = msg2["content"]
+
+        return {
+            "role": msg1["role"],
+            "content": list(content1) + list(content2),
+        }, set1 | set2
+
+    converted: List[Tuple[ConverseMessage, Set[int]]] = [
+        (await to_converse_message(msg, storage), set([idx]))
+        for idx, msg in enumerate(messages)
+        if msg.role != DialRole.SYSTEM
```

Comment:
This is a very surprising side-effect of this method.

Let's filter out system messages at the **same place** where we extract them from the list of messages.

Reply:
Done, wrapped method result in
```python
@dataclass
class ExtractSystemPromptResult:
    system_prompt: ConverseTextPart | None
    system_message_count: int
    modified_messages: List[DialMessage]


def extract_converse_system_prompt(
    messages: List[DialMessage],
) -> ExtractSystemPromptResult:
```